File: 3/3.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

func RC(min, max int) int {
	return min + rand.Intn(max-min)
}

func pop(list []int) []int {
	if len(list) > 0 {
		list = list[:len(list)-1]
	}
	return list
}

func post(m sync.Mutex, ch []int, quit chan string, number int, flag int) {
	m.Lock()
	if flag == 0 {
		fmt.Println("POST" + fmt.Sprintf("%d", number) + "started")
	}
	for i := 0; i < 1; i++ {
		if len(quit) > 0 {
			time.Sleep(1 * time.Second)
			m.Unlock()
			break
		}
		t := RC(1, 100)
		if len(ch) >= 100 {
			time.Sleep(1 * time.Second)
			m.Unlock()
			continue
		}
		if len(ch) <= 80 {
			time.Sleep(1 * time.Second)
			ch = append(ch, t)
			fmt.Println("POST" + fmt.Sprintf("%d", number) + "gave" + fmt.Sprintf("%d", t))
			m.Unlock()
		}
	}
}
func consumer(m sync.Mutex, ch []int, quit chan string, number int, flag int) {
	m.Lock()
	if flag == 0 {
		fmt.Println("CONSUMER" + fmt.Sprintf("%d", number) + "started")
	}
	for i := 0; i < 1; i++ {
		if len(ch) == 0 && len(quit) > 0 {
			time.Sleep(2 * time.Second)
			m.Unlock()
			break
		}
		if len(ch) > 1 && len(quit) == 0 {
			fmt.Println("CONSUMER " + fmt.Sprintf("%d", number) + " get" + fmt.Sprintf("%d", ch[len(ch)-1]))
			pop(ch)
			time.Sleep(2 * time.Second)
			m.Unlock()
		}
		if len(ch) == 0 && len(quit) == 0 {
			time.Sleep(2 * time.Second)
			m.Unlock()
			continue
		}
	}
}

func main() {
	var ch []int
	quit := make(chan string)
	flag := 0
	var m sync.Mutex

	for {
		for i := 0; i < 5; i++ {
			if i == 4 {
				go post(m, ch, quit, i+1, flag)
			}
			go post(m, ch, quit, i+1, flag)
			go consumer(m, ch, quit, i+1, flag)
			flag++
		}
	}
}

File: 3/3.py
import random
import keyboard
from time import sleep
from multiprocessing.pool import ThreadPool


# task executed in a worker thread
def post(list,quit,number):
    print(f'post {number} started\n')
    while True:
        if keyboard.is_pressed("q"):
            print("You pressed q")
            break
        if len(quit)>0:
            break
        if len(list)>=100:
            continue
        if len(list)<=80:
            t=random.randint(1,100)
            print(f'post {number} gave {t}\n')
            list.append(t)

def consumer(list,quit,number):
    print(f'consumer {number} started\n')
    while True:
        if keyboard.is_pressed("q"):
            print("You pressed q")
            break
        if len(list) == 0 and len(quit)==0:
            continue
        if len(list) == 0 and len(quit)>0:
            break
        if len(list) > 1:
            print(f'consumer {number} get {list[-1]}\n')
            list.pop()


# initialize a worker in the thread pool
def initialize_worker():
    # report a message
    print('Initializing worker...')


# protect the entry point
if __name__ == '__main__':
    ban = []
    qu = []
    flag=0
    # create and configure the thread pool
    with ThreadPool(5, initializer=initialize_worker) as pool:
        # issue tasks to the thread pool
        for _ in range(3):
            if _==2:
                _ = pool.apply_async(post, args=(ban, qu, flag + 1))
            _ = pool.apply_async(post,args=(ban,qu,flag+1))
            _ = pool.apply_async(consumer,args=(ban,qu,flag+1))
            flag+=1
        # close the thread pool
        pool.close()
        # wait for all tasks to complete
        pool.join()
